metricbeat/module/system/cpu/metrics/metrics.go

```diff
@@ -0,0 +1,249 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package metrics
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+)
+
+// CPU manages the CPU metrics from /proc/stat
+// If a given metric isn't available on a given platform,
+// The value will be null. All methods that use these fields
+// should assume that any value can be null.
+// The values are in "ticks", which translates to milliseconds of CPU time
+type CPU struct {
+	user    *uint64 `metric:"user"`
+	sys     *uint64 `metric:"system"`
+	idle    *uint64 `metric:"idle"`
+	nice    *uint64 `metric:"nice"`    // Linux, Darwin, BSD
+	irq     *uint64 `metric:"irq"`     // Linux and openbsd
+	wait    *uint64 `metric:"iowait"`  // Linux and AIX
+	softIrq *uint64 `metric:"softirq"` // Linux only
+	stolen  *uint64 `metric:"steal"`   // Linux only
+}
+
+// CPUMetrics carries global and per-core CPU metrics
+type CPUMetrics struct {
+	totals CPU
+	// list carries the same data, broken down by CPU
+	list []CPU
+}
+
+// Total returns the total CPU time in ticks as scraped by the API
+func (cpu CPU) Total() uint64 {
+
+	var total uint64
+	fn := func(field reflect.Value, _ int, _ string) {
+		total = total + field.Uint()
+	}
+	cpu.iterateCPUWithFunc(fn)
+	return total
+
+}
+
+// iterateCPUWithFunc uses reflection to interate over the CPU struct, stopping at every non-null field
+// `field` is the value of the struct field, iter is the field's place in the struct, and name is the `metric` tag.
+func (cpu CPU) iterateCPUWithFunc(iterFunc func(field reflect.Value, iter int, name string)) {
+	valueOfCPU := reflect.ValueOf(cpu)
+	typeOfCPU := valueOfCPU.Type()
```

Comment:
Interesting!

Reply:
> add this missing syscall to go-windows and use it here? I am ok with this being done as a follow up.

Yep. That'll be a follow-up, as deprecating gosigar is a little down the priority list.